src/Halogen/Portal.purs

```diff
@@ -44,46 +52,79 @@ type State query input output
 -- | ```purs
 -- | -- if `Nothing` is provided as the target HTMLElement, then the `<body>`
 -- | -- tag will be used
--- | HH.div_ 
--- |   [ portal _modal unit Modal.component modalInput (Just element) handler ]
+-- | HH.div_
+-- |   [ portal ntIdentity _modal unit Modal.component modalInput (Just element) handler ]
 -- |
--- | -- for comparison, this is how you would mount the component _not_ via 
+-- | -- for comparison, this is how you would mount the component _not_ via
 -- | -- a portal
 -- | HH.div_
 -- |   [ HH.slot _modal unit Modal.component modalInput handler ]
 -- | ```
 portal ::
-  forall query action input output slots label slot _1.
+  forall query action input output slots label slot _1 m n.
   Row.Cons label (H.Slot query output slot) _1 slots =>
   IsSymbol label =>
   Ord slot =>
+  MonadAff m =>
+  m (NT n Aff) ->
   SProxy label ->
   slot ->
-  H.Component HH.HTML query input output Aff ->
+  H.Component HH.HTML query input output n ->
   input ->
   Maybe HTMLElement ->
   (output -> Maybe action) ->
-  H.ComponentHTML action slots Aff
-portal label slot childComponent childInput htmlElement handler =
+  H.ComponentHTML action slots m
+portal contextualize label slot childComponent childInput htmlElement handler =
   handler
-    # HH.slot label slot component
+    # HH.slot label slot (component contextualize)
         { child: childComponent
         , input: childInput
         , targetElement: htmlElement
         }
 
+-- | Run a portal component that is already in `Aff`.
+portalAff ::
+  forall query action input output slots label slot _1.
+  Row.Cons label (H.Slot query output slot) _1 slots =>
+  IsSymbol label =>
+  Ord slot =>
+  SProxy label ->
+  slot ->
+  H.Component HH.HTML query input output Aff ->
+  input ->
+  Maybe HTMLElement ->
+  (output -> Maybe action) ->
+  H.ComponentHTML action slots Aff
+portalAff = portal (pure ntIdentity)
+
+-- | Run a portal component that is in `ReaderT r Aff` (for some context `r`).
```

Comment:
Let's say you have some app monad, `AppM`, like the [one in Real World Halogen](https://github.com/thomashoneyman/purescript-halogen-realworld/blob/f38406eefc6d094df8a4b070aacf161bc5ed8d5d/src/AppM.purs#L87).

```purs
newtype AppM a = AppM (ReaderT Env Aff a)
```

which can be translated to `Aff` with:

```purs
runAppM :: Env -> AppM ~> Aff
runAppM env (AppM m) = runReaderT m env
```

but it'd be terrible (impossible?) to thread the environment all the way down layers of your application to whatever is using the portal.

I might just be blanking, but I'm not sure how to take a component running in say `AppM` and translate it to `ReaderT Env Aff` in order to use `portalReaderT`. Do you know how to do this?

Reply:
Ah, that's a good use case to cover too. You should be able to use `H.hoist`, like we do below, to unwrap the `AppM`: `H.hoist (\(AppM ma) -> ma) child`. Or you could make your own `NT`, it amounts to the same thing: `AppM $ asks \r -> NT \(AppM ma) -> runReaderT ma r`.